File: cs/Basic/Server.HDA.Sql/SampleHistorian.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace HDA.Sql
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    using Opc.UaFx;
    using Opc.UaFx.Server;

    /// <summary>
    /// Represents a sample implementation of a custom IOpcNodeHistoryProvider.
    /// </summary>
    internal class SampleHistorian : IOpcNodeHistoryProvider
    {
        #region ---------- Private readonly fields ----------

        private readonly object syncRoot;

        #endregion

        #region ---------- Private fields ----------

        private bool autoUpdateHistory;

        #endregion

        #region ---------- Public constructors ----------

        public SampleHistorian(OpcNodeManager owner, OpcVariableNode node)
            : base()
        {
            this.Owner = owner;
            this.Node = node;

            this.Node.AccessLevel |= OpcAccessLevel.HistoryReadOrWrite;
            this.Node.UserAccessLevel |= OpcAccessLevel.HistoryReadOrWrite;
            this.Node.IsHistorizing = true;

            this.syncRoot = new object();

            this.History = SampleHistory<OpcHistoryValue>
                    .Create(@".\History.db", node.Id);
            this.ModifiedHistory = SampleHistory<OpcModifiedHistoryValue>
                    .Create(@".\History.Modified.db", node.Id);
        }

        #endregion

        #region ---------- Public properties ----------

        public bool AutoUpdateHistory
        {
            get
            {
                lock (this.syncRoot)
                    return this.autoUpdateHistory;
            }

            set
            {
                lock (this.syncRoot) {
                    if (this.autoUpdateHistory != value) {
                        this.autoUpdateHistory = value;

                        if (this.autoUpdateHistory) {
                            this.Node.BeforeApplyChanges
                                    += this.HandleNodeBeforeApplyChanges;
                        }
                        else {
                            this.Node.BeforeApplyChanges
                                    -= this.HandleNodeBeforeApplyChanges;
                        }
                    }
                }
            }
        }

        public SampleHistory<OpcHistoryValue> History
        {
            get;
        }

        public SampleHistory<OpcModifiedHistoryValue> ModifiedHistory
        {
            get;
        }

        public OpcVariableNode Node
        {
            get;
        }

        public OpcNodeManager Owner
        {
            get;
        }

        #endregion

        #region ---------- Public methods ----------

        public OpcStatusCollection CreateHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                OpcValueCollection values)
        {
            var results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count);

            lock (this.syncRoot) {
                var expectedDataType = this.Node.DataTypeId;

                for (int index = 0; index < values.Count; index++) {
                    var result = results[index];
                    var value = OpcHistoryValue.Create(values[index]);

                    if (value.DataTypeId == expectedDataType) {
                        if (this.History.Contains(value.Timestamp)) {
                            result.Update(OpcStatusCode.BadEntryExists);
                        }
                        else {
                            this.History.Add(value);

                            var modifiedValue = value.CreateModified(modificationInfo);
                            this.ModifiedHistory.Add(modifiedValue);

                            result.Update(OpcStatusCode.GoodEntryInserted);
                        }
                    }
                    else {
                        result.Update(OpcStatusCode.BadTypeMismatch);
                    }
                }
            }

            return results;
        }

        public OpcStatusCollection DeleteHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                IEnumerable<DateTime> times)
        {
            var results = OpcStatusCollection.Create(OpcStatusCode.Good, times.Count());

            lock (this.syncRoot) {
                int index = 0;

                foreach (var time in times) {
                    var result = results[index++];

                    if (this.History.Contains(time)) {
                        var value = this.History[time];
                        this.History.RemoveAt(time);

                        var modifiedValue = value.CreateModified(modificationInfo);
                        this.ModifiedHistory.Add(modifiedValue);
                    }
                    else {
                        result.Update(OpcStatusCode.BadNoEntryExists);
                    }
                }
            }

            return results;
        }

        public OpcStatusCollection DeleteHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                OpcValueCollection values)
        {
            var results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count);

            lock (this.syncRoot) {
                for (int index = 0; index < values.Count; index++) {
                    var timestamp = OpcHistoryValue.Create(values[index]).Timestamp;
                    var result = results[index];

                    if (this.History.Contains(timestamp)) {
                        var value = this.History[timestamp];
                        this.History.RemoveAt(timestamp);

                        var modifiedValue = value.CreateModified(modificationInfo);
                        this.ModifiedHistory.Add(modifiedValue);
                    }
                    else {
                        result.Update(OpcStatusCode.BadNoEntryExists);
                    }
                }
            }

            return results;
        }

        public OpcStatusCollection DeleteHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                DateTime? startTime,
                DateTime? endTime,
                OpcDeleteHistoryOptions options)
        {
            var results = new OpcStatusCollection();

            lock (this.syncRoot) {
                if (options.HasFlag(OpcDeleteHistoryOptions.Modified)) {
                    this.ModifiedHistory.RemoveRange(startTime, endTime);
                }
                else {
                    var values = this.History.Enumerate(startTime, endTime).ToArray();
                    this.History.RemoveRange(startTime, endTime);

                    for (int index = 0; index < values.Length; index++) {
                        var value = values[index];
                        this.ModifiedHistory.Add(value.CreateModified(modificationInfo));

                        results.Add(OpcStatusCode.Good);
                    }
                }
            }

            return results;
        }

        public IEnumerable<OpcHistoryValue> ReadHistory(
                OpcContext context,
                DateTime? startTime,
                DateTime? endTime,
                OpcReadHistoryOptions options)
        {
            lock (this.syncRoot) {
                if (options.HasFlag(OpcReadHistoryOptions.Modified)) {
                    return this.ModifiedHistory
                            .Enumerate(startTime, endTime)
                            .Cast<OpcHistoryValue>()
                            .ToArray();
                }

                return this.History
                        .Enumerate(startTime, endTime)
                        .ToArray();
            }
        }

        public OpcStatusCollection ReplaceHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                OpcValueCollection values)
        {
            var results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count);

            lock (this.syncRoot) {
                var expectedDataTypeId = this.Node.DataTypeId;

                for (int index = 0; index < values.Count; index++) {
                    var result = results[index];
                    var value = OpcHistoryValue.Create(values[index]);

                    if (value.DataTypeId == expectedDataTypeId) {
                        if (this.History.Contains(value.Timestamp)) {
                            var oldValue = this.History[value.Timestamp];
                            this.History.Replace(value);

                            var modifiedValue = oldValue.CreateModified(modificationInfo);
                            this.ModifiedHistory.Add(modifiedValue);

                            result.Update(OpcStatusCode.GoodEntryReplaced);
                        }
                        else {
                            result.Update(OpcStatusCode.BadNoEntryExists);
                        }
                    }
                    else {
                        result.Update(OpcStatusCode.BadTypeMismatch);
                    }
                }
            }

            return results;
        }

        public OpcStatusCollection UpdateHistory(
                OpcContext context,
                OpcHistoryModificationInfo modificationInfo,
                OpcValueCollection values)
        {
            var results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count);

            lock (this.syncRoot) {
                var expectedDataTypeId = this.Node.DataTypeId;

                for (int index = 0; index < values.Count; index++) {
                    var result = results[index];
                    var value = OpcHistoryValue.Create(values[index]);

                    if (value.DataTypeId == expectedDataTypeId) {
                        if (this.History.Contains(value.Timestamp)) {
                            var oldValue = this.History[value.Timestamp];
                            this.History.Replace(value);

                            var modifiedValue = oldValue.CreateModified(modificationInfo);
                            this.ModifiedHistory.Add(modifiedValue);

                            result.Update(OpcStatusCode.GoodEntryReplaced);
                        }
                        else {
                            this.History.Add(value);

                            var modifiedValue = value.CreateModified(modificationInfo);
                            this.ModifiedHistory.Add(modifiedValue);

                            result.Update(OpcStatusCode.GoodEntryInserted);
                        }
                    }
                    else {
                        result.Update(OpcStatusCode.BadTypeMismatch);
                    }
                }
            }

            return results;
        }

        #endregion

        #region ---------- Private methods ----------

        private void HandleNodeBeforeApplyChanges(object sender, OpcNodeChangesEventArgs e)
        {
            var timestamp = this.Node.Timestamp;

            if (timestamp != null && e.IsChangeOf(OpcNodeChanges.Value)) {
                var value = new OpcHistoryValue(this.Node.Value, timestamp.Value);

                if (this.History.Contains(value.Timestamp))
                    this.History.Replace(value);
                else
                    this.History.Add(value);
            }
        }

        #endregion
    }
}

File: vb/Basic/Server.HDA.Sql/SampleHistorian.vb
' Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports System
Imports System.Collections.Generic
Imports System.Linq

Imports Opc.UaFx
Imports Opc.UaFx.Server

Namespace HDA.Sql
    ''' <summary>
    ''' Represents a sample implementation of a custom IOpcNodeHistoryProvider.
    ''' </summary>
    Friend Class SampleHistorian
        Implements IOpcNodeHistoryProvider

        '---------- Private readonly fields ----------

        Private ReadOnly syncRoot As Object

        '---------- Private fields ----------

        Private _autoUpdateHistory As Boolean
        Private _node As OpcVariableNode

        '---------- Public constructors ----------

        Public Sub New(ByVal owner As OpcNodeManager, ByVal node As OpcVariableNode)
            MyBase.New()
            Me.Owner = owner
            Me._node = node

            Me._node.AccessLevel = Me._node.AccessLevel Or OpcAccessLevel.HistoryReadOrWrite
            Me._node.UserAccessLevel = Me._node.UserAccessLevel Or OpcAccessLevel.HistoryReadOrWrite
            Me._node.IsHistorizing = True

            Me.syncRoot = New Object()

            Me.History = SampleHistory(Of OpcHistoryValue) _
                    .Create(".\History.db", node.Id)
            Me.ModifiedHistory = SampleHistory(Of OpcModifiedHistoryValue) _
                    .Create(".\History.Modified.db", node.Id)
        End Sub

        '---------- Public properties ----------

        Public Property AutoUpdateHistory As Boolean
            Get
                SyncLock Me.syncRoot
                    Return Me._autoUpdateHistory
                End SyncLock
            End Get

            Set(ByVal value As Boolean)
                SyncLock Me.syncRoot
                    If Me._autoUpdateHistory <> value Then
                        Me._autoUpdateHistory = value

                        If Me._autoUpdateHistory Then
                            AddHandler Me._node.BeforeApplyChanges, AddressOf Me.HandleNodeBeforeApplyChanges
                        Else
                            RemoveHandler Me._node.BeforeApplyChanges, AddressOf Me.HandleNodeBeforeApplyChanges
                        End If
                    End If
                End SyncLock
            End Set
        End Property

        Public ReadOnly Property History As SampleHistory(Of OpcHistoryValue)
        Public ReadOnly Property ModifiedHistory As SampleHistory(Of OpcModifiedHistoryValue)

        Public ReadOnly Property Node As OpcVariableNode
            Get
                Return Me._node
            End Get
        End Property

        Public ReadOnly Property Owner As OpcNodeManager

        '---------- Private properties (by IOpcNodeHistoryProvider) ----------

        Private ReadOnly Property IOpcNodeHistoryProvider_Node As OpcVariableNode Implements IOpcNodeHistoryProvider.Node
            Get
                Return Me._node
            End Get
        End Property

        '---------- Private methods (by IOpcNodeHistoryProvider) ----------

        Private Function IOpcNodeHistoryProvider_CreateHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                values As OpcValueCollection) As OpcStatusCollection Implements IOpcNodeHistoryProvider.CreateHistory
            Dim results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count)

            SyncLock Me.syncRoot
                Dim expectedDataType = Me._node.DataTypeId

                For index As Integer = 0 To values.Count - 1
                    Dim result = results(index)
                    Dim value = OpcHistoryValue.Create(values(index))

                    If value.DataTypeId = expectedDataType Then
                        If Me.History.Contains(value.Timestamp) Then
                            result.Update(OpcStatusCode.BadEntryExists)
                        Else
                            Me.History.Add(value)

                            Dim modifiedValue = value.CreateModified(modificationInfo)
                            Me.ModifiedHistory.Add(modifiedValue)

                            result.Update(OpcStatusCode.GoodEntryInserted)
                        End If
                    Else
                        result.Update(OpcStatusCode.BadTypeMismatch)
                    End If
                Next
            End SyncLock

            Return results
        End Function

        Private Function IOpcNodeHistoryProvider_DeleteHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                times As IEnumerable(Of DateTime)) As OpcStatusCollection Implements IOpcNodeHistoryProvider.DeleteHistory
            Dim results = OpcStatusCollection.Create(OpcStatusCode.Good, times.Count())

            SyncLock Me.syncRoot
                Dim index As Integer = 0

                For Each time In times
                    index += 1
                    Dim result = results(index)

                    If Me.History.Contains(time) Then
                        Dim value = Me.History(time)
                        Me.History.RemoveAt(time)

                        Dim modifiedValue = value.CreateModified(modificationInfo)
                        Me.ModifiedHistory.Add(modifiedValue)
                    Else
                        result.Update(OpcStatusCode.BadNoEntryExists)
                    End If
                Next
            End SyncLock

            Return results
        End Function

        Private Function IOpcNodeHistoryProvider_DeleteHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                values As OpcValueCollection) As OpcStatusCollection Implements IOpcNodeHistoryProvider.DeleteHistory
            Dim results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count)

            SyncLock Me.syncRoot
                For index As Integer = 0 To values.Count - 1
                    Dim timestamp = OpcHistoryValue.Create(values(index)).Timestamp
                    Dim result = results(index)

                    If Me.History.Contains(timestamp) Then
                        Dim value = Me.History(timestamp)
                        Me.History.RemoveAt(timestamp)

                        Dim modifiedValue = value.CreateModified(modificationInfo)
                        Me.ModifiedHistory.Add(modifiedValue)
                    Else
                        result.Update(OpcStatusCode.BadNoEntryExists)
                    End If
                Next
            End SyncLock

            Return results
        End Function

        Private Function IOpcNodeHistoryProvider_DeleteHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                startTime As DateTime?,
                endTime As DateTime?,
                options As OpcDeleteHistoryOptions) As OpcStatusCollection Implements IOpcNodeHistoryProvider.DeleteHistory
            Dim results = New OpcStatusCollection()

            SyncLock Me.syncRoot
                If options.HasFlag(OpcDeleteHistoryOptions.Modified) Then
                    Me.ModifiedHistory.RemoveRange(startTime, endTime)
                Else
                    Dim values = Me.History.Enumerate(startTime, endTime).ToArray()
                    Me.History.RemoveRange(startTime, endTime)

                    For index As Integer = 0 To values.Length - 1
                        Dim value = values(index)
                        Me.ModifiedHistory.Add(value.CreateModified(modificationInfo))

                        results.Add(OpcStatusCode.Good)
                    Next
                End If
            End SyncLock

            Return results
        End Function

        Private Function IOpcNodeHistoryProvider_ReadHistory(
                context As OpcContext,
                startTime As DateTime?,
                endTime As DateTime?,
                timeFlowsBackward As Boolean,
                maxCount As UInteger?,
                options As OpcReadHistoryOptions) As IEnumerable(Of OpcHistoryValue) Implements IOpcNodeHistoryProvider.ReadHistory
            Dim historyValues As OpcHistoryValue()

            SyncLock Me.syncRoot
                If options.HasFlag(OpcReadHistoryOptions.Modified) Then
                    historyValues = Me.ModifiedHistory _
                            .Enumerate(startTime, endTime) _
                            .Cast(Of OpcHistoryValue)() _
                            .ToArray()
                End If

                historyValues = Me.History _
                        .Enumerate(startTime, endTime) _
                        .ToArray()
            End SyncLock

            ' If timeFlowsBackward Is true, we need to reverse the array in order to
            ' return the values in descending order.
            If timeFlowsBackward Then
                Array.Reverse(historyValues)
            End If

            Return historyValues
        End Function

        Private Function IOpcNodeHistoryProvider_ReplaceHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                values As OpcValueCollection) As OpcStatusCollection Implements IOpcNodeHistoryProvider.ReplaceHistory
            Dim results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count)

            SyncLock Me.syncRoot
                Dim expectedDataTypeId = Me._node.DataTypeId

                For index As Integer = 0 To values.Count - 1
                    Dim result = results(index)
                    Dim value = OpcHistoryValue.Create(values(index))

                    If value.DataTypeId = expectedDataTypeId Then

                        If Me.History.Contains(value.Timestamp) Then
                            Dim oldValue = Me.History(value.Timestamp)
                            Me.History.Replace(value)

                            Dim modifiedValue = oldValue.CreateModified(modificationInfo)
                            Me.ModifiedHistory.Add(modifiedValue)

                            result.Update(OpcStatusCode.GoodEntryReplaced)
                        Else
                            result.Update(OpcStatusCode.BadNoEntryExists)
                        End If
                    Else
                        result.Update(OpcStatusCode.BadTypeMismatch)
                    End If
                Next
            End SyncLock

            Return results
        End Function

        Private Function IOpcNodeHistoryProvider_UpdateHistory(
                context As OpcContext,
                modificationInfo As OpcHistoryModificationInfo,
                values As OpcValueCollection) As OpcStatusCollection Implements IOpcNodeHistoryProvider.UpdateHistory
            Dim results = OpcStatusCollection.Create(OpcStatusCode.Good, values.Count)

            SyncLock Me.syncRoot
                Dim expectedDataTypeId = Me._node.DataTypeId

                For index As Integer = 0 To values.Count - 1
                    Dim result = results(index)
                    Dim value = OpcHistoryValue.Create(values(index))

                    If value.DataTypeId = expectedDataTypeId Then
                        If Me.History.Contains(value.Timestamp) Then
                            Dim oldValue = Me.History(value.Timestamp)
                            Me.History.Replace(value)

                            Dim modifiedValue = oldValue.CreateModified(modificationInfo)
                            Me.ModifiedHistory.Add(modifiedValue)

                            result.Update(OpcStatusCode.GoodEntryReplaced)
                        Else
                            Me.History.Add(value)

                            Dim modifiedValue = value.CreateModified(modificationInfo)
                            Me.ModifiedHistory.Add(modifiedValue)

                            result.Update(OpcStatusCode.GoodEntryInserted)
                        End If
                    Else
                        result.Update(OpcStatusCode.BadTypeMismatch)
                    End If
                Next
            End SyncLock

            Return results
        End Function

        '---------- Private methods ----------

        Private Sub HandleNodeBeforeApplyChanges(ByVal sender As Object, ByVal e As OpcNodeChangesEventArgs)
            Dim timestamp = Me._node.Timestamp

            If timestamp IsNot Nothing AndAlso e.IsChangeOf(OpcNodeChanges.Value) Then
                Dim value = New OpcHistoryValue(Me._node.Value, timestamp.Value)

                If Me.History.Contains(value.Timestamp) Then
                    Me.History.Replace(value)
                Else
                    Me.History.Add(value)
                End If
            End If
        End Sub
    End Class
End Namespace
